File: branches/guava/jung/jung-visualization/src/main/java/edu/uci/ics/jung/visualization/annotations/AnnotationManager.java
/*
 * Copyright (c) 2005, the JUNG Project and the Regents of the University of
 * California All rights reserved.
 *
 * This software is open-source under the BSD license; see either "license.txt"
 * or http://jung.sourceforge.net/license.txt for a description.
 *
 * 
 */
package edu.uci.ics.jung.visualization.annotations;

import java.awt.Component;
import java.awt.Dimension;
import java.awt.Shape;
import java.awt.geom.AffineTransform;
import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

import edu.uci.ics.jung.visualization.Layer;
import edu.uci.ics.jung.visualization.RenderContext;
import edu.uci.ics.jung.visualization.transform.AffineTransformer;
import edu.uci.ics.jung.visualization.transform.LensTransformer;
import edu.uci.ics.jung.visualization.transform.MutableTransformer;

/**
 * handles the selection of annotations, and the support for the
 * tools to draw them at specific layers.
 * 
 * @author Tom Nelson - tomnelson@dev.java.net
 *
 */
public class AnnotationManager {
	
    protected AnnotationRenderer annotationRenderer = new AnnotationRenderer();
	protected AnnotationPaintable lowerAnnotationPaintable;
	protected AnnotationPaintable upperAnnotationPaintable;
	
	protected RenderContext<?,?> rc;
	protected AffineTransformer transformer;

	public AnnotationManager(RenderContext<?,?> rc) {
		this.rc = rc;
		this.lowerAnnotationPaintable = new AnnotationPaintable(rc, annotationRenderer);
		this.upperAnnotationPaintable = new AnnotationPaintable(rc, annotationRenderer);
		
		MutableTransformer mt = rc.getMultiLayerTransformer().getTransformer(Layer.LAYOUT);
		if(mt instanceof AffineTransformer) {
			transformer = (AffineTransformer)mt;
		} else if(mt instanceof LensTransformer) {
			transformer = (AffineTransformer)((LensTransformer)mt).getDelegate();
		}

	}
	
	public AnnotationPaintable getAnnotationPaintable(Annotation.Layer layer) {
		if(layer == Annotation.Layer.LOWER) {
			return this.lowerAnnotationPaintable;
		}
		if(layer == Annotation.Layer.UPPER) {
			return this.upperAnnotationPaintable;
		}
		return null;
	}
	
	public void add(Annotation.Layer layer, Annotation<?> annotation) {
		if(layer == Annotation.Layer.LOWER) {
			this.lowerAnnotationPaintable.add(annotation);
		}
		if(layer == Annotation.Layer.UPPER) {
			this.upperAnnotationPaintable.add(annotation);
		}
	}
	
	public void remove(Annotation<?> annotation) {
		this.lowerAnnotationPaintable.remove(annotation);
		this.upperAnnotationPaintable.remove(annotation);
	}
	
	protected AnnotationPaintable getLowerAnnotationPaintable() {
		return lowerAnnotationPaintable;
	}
	
	protected AnnotationPaintable getUpperAnnotationPaintable() {
		return upperAnnotationPaintable;
	}
	
	@SuppressWarnings("unchecked")
    public Annotation getAnnotation(Point2D p) {
		Set<Annotation> annotations = new HashSet<Annotation>(lowerAnnotationPaintable.getAnnotations());
		annotations.addAll(upperAnnotationPaintable.getAnnotations());
		return getAnnotation(p, annotations);
	}
	
	@SuppressWarnings("unchecked")
    public Annotation getAnnotation(Point2D p, Collection<Annotation> annotations) {
		double closestDistance = Double.MAX_VALUE;
		Annotation closestAnnotation = null;
		for(Annotation annotation : annotations) {
			Object ann = annotation.getAnnotation();
			if(ann instanceof Shape) {
				Point2D ip = rc.getMultiLayerTransformer().inverseTransform(p);
				Shape shape = (Shape)ann;
				if(shape.contains(ip)) {
					
					Rectangle2D shapeBounds = shape.getBounds2D();
					Point2D shapeCenter = new Point2D.Double(shapeBounds.getCenterX(), shapeBounds.getCenterY());
					double distanceSq = shapeCenter.distanceSq(ip);
					if(distanceSq < closestDistance) {
						closestDistance = distanceSq;
						closestAnnotation = annotation;
					}
				}
			} else if(ann instanceof String) {
				
				Point2D ip = rc.getMultiLayerTransformer().inverseTransform(Layer.VIEW, p);
				Point2D ap = annotation.getLocation();
				String label = (String)ann;
				Component component = prepareRenderer(rc, annotationRenderer, label);
				
				AffineTransform base = new AffineTransform(transformer.getTransform());
				double rotation = transformer.getRotation();
				// unrotate the annotation
				AffineTransform unrotate =
					AffineTransform.getRotateInstance(-rotation, ap.getX(), ap.getY());
				base.concatenate(unrotate);
				
				Dimension d = component.getPreferredSize();
				Rectangle2D componentBounds = new Rectangle2D.Double(ap.getX(), ap.getY(), d.width, d.height);
				
				Shape componentBoundsShape = base.createTransformedShape(componentBounds);
				Point2D componentCenter = new Point2D.Double(componentBoundsShape.getBounds().getCenterX(),
						componentBoundsShape.getBounds().getCenterY());
				if(componentBoundsShape.contains(ip)) {
					double distanceSq = componentCenter.distanceSq(ip);
					if(distanceSq < closestDistance) {
						closestDistance = distanceSq;
						closestAnnotation = annotation;
					}
				}
				
			}
		}
		return closestAnnotation;
	}
	
	public Component prepareRenderer(RenderContext<?,?> rc, AnnotationRenderer annotationRenderer, Object value) {
		return annotationRenderer.getAnnotationRendererComponent(rc.getScreenDevice(), value);
	}


	
	

}

File: jung-visualization/src/main/kotlin/edu/uci/ics/jung/visualization/annotations/AnnotationManager.kt
/*
 * Copyright (c) 2005, The JUNG Authors
 * All rights reserved.
 *
 * This software is open-source under the BSD license; see either "license.txt"
 * or https://github.com/jrtom/jung/blob/master/LICENSE for a description.
 */
package edu.uci.ics.jung.visualization.annotations

import edu.uci.ics.jung.visualization.MultiLayerTransformer
import edu.uci.ics.jung.visualization.RenderContext
import edu.uci.ics.jung.visualization.transform.AffineTransformer
import edu.uci.ics.jung.visualization.transform.LensTransformer
import java.awt.Component
import java.awt.Dimension
import java.awt.Shape
import java.awt.geom.AffineTransform
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D

/**
 * handles the selection of annotations, and the support for the tools to draw them at specific
 * layers.
 *
 * @author Tom Nelson - tomnelson@dev.java.net
 */
open class AnnotationManager(private val rc: RenderContext<*, *>) {

    private val annotationRenderer: AnnotationRenderer = AnnotationRenderer()
    val lowerAnnotationPaintable: AnnotationPaintable = AnnotationPaintable(rc, annotationRenderer)
    val upperAnnotationPaintable: AnnotationPaintable = AnnotationPaintable(rc, annotationRenderer)
    private val transformer: AffineTransformer

    init {
        val mt = rc.getMultiLayerTransformer().getTransformer(MultiLayerTransformer.Layer.LAYOUT)
        transformer = when (mt) {
            is AffineTransformer -> mt
            is LensTransformer -> mt.delegate as AffineTransformer
            else -> throw IllegalStateException("Unexpected transformer type: ${mt::class.java}")
        }
    }

    fun getAnnotationPaintable(layer: Annotation.Layer): AnnotationPaintable? {
        return when (layer) {
            Annotation.Layer.LOWER -> lowerAnnotationPaintable
            Annotation.Layer.UPPER -> upperAnnotationPaintable
        }
    }

    fun add(layer: Annotation.Layer, annotation: Annotation<*>) {
        when (layer) {
            Annotation.Layer.LOWER -> lowerAnnotationPaintable.add(annotation)
            Annotation.Layer.UPPER -> upperAnnotationPaintable.add(annotation)
        }
    }

    fun remove(annotation: Annotation<*>?) {
        if (annotation == null) return
        lowerAnnotationPaintable.remove(annotation)
        upperAnnotationPaintable.remove(annotation)
    }

    fun getAnnotation(p: Point2D): Annotation<*>? {
        val annotations = HashSet<Annotation<*>>(lowerAnnotationPaintable.getAnnotations())
        annotations.addAll(upperAnnotationPaintable.getAnnotations())
        return getAnnotation(p, annotations)
    }

    fun getAnnotation(p: Point2D, annotations: Collection<Annotation<*>>): Annotation<*>? {
        var closestDistance = Double.MAX_VALUE
        var closestAnnotation: Annotation<*>? = null
        for (annotation in annotations) {
            val ann = annotation.annotation
            if (ann is Shape) {
                val ip = rc.getMultiLayerTransformer().inverseTransform(p)
                val shape = ann
                if (shape.contains(ip)) {
                    val shapeBounds = shape.bounds2D
                    val shapeCenter = Point2D.Double(shapeBounds.centerX, shapeBounds.centerY)
                    val distanceSq = shapeCenter.distanceSq(ip)
                    if (distanceSq < closestDistance) {
                        closestDistance = distanceSq
                        closestAnnotation = annotation
                    }
                }
            } else if (ann is String) {
                val ip = rc.getMultiLayerTransformer()
                    .inverseTransform(MultiLayerTransformer.Layer.VIEW, p)
                val ap = annotation.location
                val label = ann
                val component = prepareRenderer(rc, annotationRenderer, label)

                val base = AffineTransform(transformer.getTransform())
                val rotation = transformer.getRotation()
                // unrotate the annotation
                val unrotate = AffineTransform.getRotateInstance(-rotation, ap.x, ap.y)
                base.concatenate(unrotate)

                val d = component.preferredSize
                val componentBounds = Rectangle2D.Double(
                    ap.x, ap.y, d.width.toDouble(), d.height.toDouble()
                )

                val componentBoundsShape = base.createTransformedShape(componentBounds)
                val componentCenter = Point2D.Double(
                    componentBoundsShape.bounds.centerX,
                    componentBoundsShape.bounds.centerY
                )
                if (componentBoundsShape.contains(ip)) {
                    val distanceSq = componentCenter.distanceSq(ip)
                    if (distanceSq < closestDistance) {
                        closestDistance = distanceSq
                        closestAnnotation = annotation
                    }
                }
            }
        }
        return closestAnnotation
    }

    fun prepareRenderer(
        rc: RenderContext<*, *>, annotationRenderer: AnnotationRenderer, value: Any?
    ): Component =
        annotationRenderer.getAnnotationRendererComponent(rc.getScreenDevice(), value)
}
